src/main/java/com/linecorp/armeria/server/metrics/ServiceMetricConsumer.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2015 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.metrics;
+
+import java.util.Objects;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.linecorp.armeria.common.SessionProtocol;
+
+@FunctionalInterface
+public interface ServiceMetricConsumer {
+    Logger logger = LoggerFactory.getLogger(ServiceMetricConsumer.class);
+
+    /**
+     * Invoked for each request that has been processed
+     *
+     * @param sessionProtocol session protocol, currently only http-based session protocols are supported.
+     * @param code the http status code, as defined in
+     *             <a href="https://tools.ietf.org/html/rfc7231#section-6">the section 6 of RFC7231</a>
+     * @param processTimeNanos elapsed nano time processing request
+     * @param requestSize number of bytes in request if possible, otherwise it will be 0
+     * @param responseSize number of bytes in response if possible, otherwise it will be 0
+     */
+    void invocationComplete(SessionProtocol sessionProtocol, int code, long processTimeNanos, int requestSize,
```

Comment:
I wonder if it's a good idea to split this method into two:

- `invocationComplete()` - invoked when `DecodeResultType` is `SUCCESS`
  - Replace `sessionProtocol` with `ServiceInvocationContext`
- `codecFailed()` invoked when `DecodeResultType` is not `SUCCESS`
  - Add `hostname`, `path`, and `mappedPath` as parameters

.. so that the consumer has access to more information about the invocation (path, hostname, etc).